timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -505,25 +423,73 @@ void __global__ k_nonbonded_unified(
     unsigned long long *__restrict__ du_dx,
     unsigned long long *__restrict__ du_dp,
     unsigned long long *__restrict__ u_buffer) {
-
-    int tile_idx = blockIdx.x;
-    int row_block_idx = ixn_tiles[tile_idx];
-    int index = row_block_idx * 32 + threadIdx.x;
-    const unsigned int atom_i_idx = index < NR ? row_idxs[index] : N;
-    int atom_j_idx = ixn_atoms[tile_idx * 32 + threadIdx.x];
-
-    RealType w_i = atom_i_idx < N ? params[atom_i_idx * PARAMS_PER_ATOM + PARAM_OFFSET_W] : 0;
-    RealType w_j = atom_j_idx < N ? params[atom_j_idx * PARAMS_PER_ATOM + PARAM_OFFSET_W] : 0;
-
-    int is_vanilla = w_i == 0 && w_j == 0;
-
-    bool tile_is_vanilla = __all_sync(0xffffffff, is_vanilla);
-
-    if (tile_is_vanilla) {
-        v_nonbonded_unified<RealType, 0, COMPUTE_U, COMPUTE_DU_DX, COMPUTE_DU_DP>(
-            N, NR, coords, params, box, beta, cutoff, row_idxs, ixn_tiles, ixn_atoms, du_dx, du_dp, u_buffer);
-    } else {
-        v_nonbonded_unified<RealType, 1, COMPUTE_U, COMPUTE_DU_DX, COMPUTE_DU_DP>(
-            N, NR, coords, params, box, beta, cutoff, row_idxs, ixn_tiles, ixn_atoms, du_dx, du_dp, u_buffer);
-    };
+    __shared__ RealType shared_box[6];
+    if (threadIdx.x == 0) {
+        shared_box[0] = box[0 * 3 + 0];
+        shared_box[1] = box[1 * 3 + 1];
+        shared_box[2] = box[2 * 3 + 2];
+        shared_box[3] = 1 / shared_box[0];
+        shared_box[4] = 1 / shared_box[1];
+        shared_box[5] = 1 / shared_box[2];
+    }
+    __syncthreads();
+    // Tiles are 32 x 32, which is the same as the warp size
+    const int tile_size = warp_size;
+
+    const int tiles_per_block = blockDim.x / warp_size;
```

Comment:
This is a bit more ambiguous as its a question of whether we want to run multiple tiles per warp or a single tile per warp regardless of tile size.

Will change `tile_offset` to use `warp_size` instead of `tile_size` to be consistent with the tiles per warp convention, but will need to be revisited for performance once we start having different sizes of tiles.

Reply:
Changed in [36d2a0b](https://github.com/proteneer/timemachine/pull/1047/commits/36d2a0bd5619037f5d62f249d5272cead028d459)